sympy/matrices/graph.py

```diff
@@ -0,0 +1,135 @@
+from sympy.utilities.iterables import \
+    flatten, strongly_connected_components
+
+from .common import NonSquareMatrixError
+
+
+def _connected_components(M):
+    """Returns the list of connected vertices of the graph when
+    a square matrix is viewed as a weighted graph.
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, Matrix
+    >>> a, b, c, d, e, f, g, h = symbols('a:h')
+    >>> A = Matrix([
+    ...     [a, 0, b, 0],
+    ...     [0, e, 0, f],
+    ...     [c, 0, d, 0],
+    ...     [0, g, 0, h]])
+    >>> A.connected_components()
+    [[0, 2], [1, 3]]
+
+    Notes
+    =====
+
+    Even if any symbolic elements of the matrix can be indeterminate
+    to be zero mathematically, this only takes the account of the
+    structural aspect of the matrix, so they will considered to be
+    nonzero.
+    """
+    from .sparse import SparseMatrix
+
+    if not M.is_square:
+        raise NonSquareMatrixError
+
+    N = M.shape[0]
+    V = list(range(N))
+
+    if isinstance(M, SparseMatrix):
+        E = set()
+        for i, j in M._smat.keys():
```

Comment:
I think that in general it would be useful if there was a way of iterating over e.g. the nonzero elements of a row of a matrix that could be used without needing to special case sparse and dense matrices (they could each use an efficient implementation of the iteration).